java-squid/src/main/java/org/sonar/java/cfg/LiveVariables.java

```diff
@@ -0,0 +1,155 @@
+
+import java.io.PrintStream;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Set;
+
+public class LiveVariables {
+
+  private final CFG cfg;
+  private final Map<CFG.Block, Set<Symbol>> in = new HashMap<>();
+  private final Map<CFG.Block, Set<Symbol>> out = new HashMap<>();
+
+  private LiveVariables(CFG cfg) {
+    this.cfg = cfg;
+  }
+
+  public Set<Symbol> getOut(CFG.Block block) {
+    return ImmutableSet.copyOf(out.get(block));
+  }
+
+  public static LiveVariables analyze(CFG cfg) {
+    LiveVariables liveVariables = new LiveVariables(cfg);
+    liveVariables.analyze();
+    return liveVariables;
+  }
+
+  private void analyze() {
+    Map<CFG.Block, Set<Symbol>> kill = new HashMap<>();
+    Map<CFG.Block, Set<Symbol>> gen = new HashMap<>();
+    for (CFG.Block block : cfg.blocks) {
+      Set<Symbol> blockKill = new HashSet<>();
+      Set<Symbol> blockGen = new HashSet<>();
+      processElements(block, blockKill, blockGen);
+      kill.put(block, blockKill);
+      gen.put(block, blockGen);
+    }
+
+    Deque<CFG.Block> workList = new LinkedList<>();
+    workList.addAll(cfg.blocks);
+    while (!workList.isEmpty()) {
+      CFG.Block block = workList.removeFirst();
+
+      Set<Symbol> blockOut = out.get(block);
+      if (blockOut == null) {
+        blockOut = new HashSet<>();
+        out.put(block, blockOut);
+      }
+      for (CFG.Block successor : block.successors()) {
+        Set<Symbol> inOfSuccessor = in.get(successor);
+        if (inOfSuccessor != null) {
+          blockOut.addAll(inOfSuccessor);
+        }
+      }
+
+      // in = gen and (out - kill)
+      Set<Symbol> newIn = new HashSet<>(gen.get(block));
+      newIn.addAll(Sets.difference(blockOut, kill.get(block)));
+
+      if (newIn.equals(in.get(block))) {
+        continue;
+      }
+      in.put(block, newIn);
+      for (CFG.Block predecessor : block.predecessors()) {
+        workList.addLast(predecessor);
+      }
+    }
+
+    // out of exit block are empty by definition.
+    if (!out.get(cfg.blocks.get(0)).isEmpty()) {
+      throw new IllegalStateException("Out of exit block should be empty");
+    }
+  }
+
+  private void processElements(CFG.Block block, Set<Symbol> blockKill, Set<Symbol> blockGen) {
+    // process elements from bottom to top
+    Set<Tree> assignmentLHS = new HashSet<>();
+    for (Tree element : block.elements) {
+      Symbol symbol;
+      switch (element.kind()) {
+        case ASSIGNMENT:
+          ExpressionTree lhs = ((AssignmentExpressionTree) element).variable();
+          if (lhs.is(Tree.Kind.IDENTIFIER)) {
+            symbol = ((IdentifierTree) lhs).symbol();
+            if (isLocalVariable(symbol)) {
+              assignmentLHS.add(lhs);
+              blockGen.remove(symbol);
+              blockKill.add(symbol);
+            }
+          }
+          break;
+        case IDENTIFIER:
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce this switch case number of lines from 7 to at most 5, for example by extracting code into methods. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1151)